File: backend/src/AgentHost.Api/Infrastructure/Paging.cs
namespace AgentHost.Api.Infrastructure;

/// <summary>
/// Pagination guards for every list endpoint. Client-supplied <c>take</c>/<c>skip</c> values are
/// never passed to SQL unclamped: an unbounded <c>?take=100000000</c> is a trivial memory/CPU
/// denial-of-service against both the API process and Postgres.
/// </summary>
public static class Paging
{
    /// <summary>Hard upper bound on any single page, regardless of what the client asks for.</summary>
    public const int MaxTake = 200;

    /// <summary>Clamps a requested page size into [1, <paramref name="max"/>].</summary>
    public static int ClampTake(int take, int max = MaxTake)
    {
        if (max > MaxTake) max = MaxTake;
        if (take < 1) return 1;
        return take > max ? max : take;
    }

    /// <summary>Floors a requested offset at 0 (a negative OFFSET is a Postgres error).</summary>
    public static int ClampSkip(int skip) => skip < 0 ? 0 : skip;
}

File: frontend/src/app/core/utils/paging.ts
/**
 * Client-side mirror of the server's `Paging` helper (backend Infrastructure/Paging.cs).
 *
 * The server clamps silently: ask for `take=1000` and you get 200 back with nothing saying so.
 * A caller that then advanced by 1000 would skip 800 rows it never saw. Clamping here keeps the
 * page size the client *thinks* it requested equal to the one it will get, so paging arithmetic
 * stays correct. Both bounds must move together — this file exists to make that obvious.
 */

/** Hard upper bound on any single page, matching `Paging.MaxTake`. */
export const MAX_TAKE = 200;

/** Clamps a requested page size into [1, max]. */
export function clampTake(take: number, max: number = MAX_TAKE): number {
  const ceiling = Math.min(max, MAX_TAKE);
  if (!Number.isFinite(take)) return ceiling;
  const size = Math.floor(take);
  if (size < 1) return 1;
  return size > ceiling ? ceiling : size;
}

/** Floors a requested offset at 0 — a negative OFFSET is a Postgres error. */
export function clampSkip(skip: number): number {
  if (!Number.isFinite(skip)) return 0;
  const offset = Math.floor(skip);
  return offset < 0 ? 0 : offset;
}
